From d1479b3466d0c827761df50d19588302cc0d4838 Mon Sep 17 00:00:00 2001
From: Robin Gong <yibin.gong@nxp.com>
Date: Thu, 9 Feb 2017 11:00:55 +0800
Subject: [PATCH 1396/5242] MLK-13849 dma: imx-sdma: fix kernel crashs when
 play Audio by HDMI

commit  f760f8d4eaf3fb58f37dd753e7adb32becb27e1a from
https://source.codeaurora.org/external/imx/linux-imx.git

In cyclic/loop mode, DMA call shouldn't touch dma cookie, since only one
time need to be submitted, otherwise, below kernel bug will be triggered:

Playing WAVE '/unit_tests/audio8k16S.wav' : Signed 16 bit Little Endian, Rate 8000 Hz, Stereo
Signed-off-by: Meng Li <Meng.Li@windriver.com>
------------[ cut here ]------------
Kernel BUG at 8043d030 [verbose debug info unavailable]
Internal error: Oops - BUG: 0 [#1] PREEMPT SMP ARM
Modules linked in: ov5642_camera mxc_v4l2_capture ipu_bg_overlay_sdc ipu_still ipu_prp_enc ipu_csi_enc ipu_fg_overlay_sdc ov5640_camera_mipi_int ov5640_camera_int v4l2_int_device mxc_dcic
CPU: 0 PID: 0 Comm: swapper/0 Not tainted 4.9.5-01681-g6dbd27b #7
Hardware name: Freescale i.MX6 Quad/DualLite (Device Tree)
task: 80f06600 task.stack: 80f00000
PC is at mxc_sdma_handle_channel_normal+0xc0/0xd8
LR is at tasklet_action+0x94/0x14c

Signed-off-by: Robin Gong <yibin.gong@nxp.com>
Signed-off-by: Mihai Serban <mihai.serban@nxp.com>
---
 drivers/dma/imx-sdma.c |   12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index 9168f39..e3534eb 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -838,10 +838,14 @@ static irqreturn_t sdma_int_handler(int irq, void *dev_id)
 		int channel = fls(stat) - 1;
 		struct sdma_channel *sdmac = &sdma->channel[channel];
 
-		if ((sdmac->flags & IMX_DMA_SG_LOOP) &&
-			(sdmac->peripheral_type != IMX_DMATYPE_HDMI))
-			sdma_update_channel_loop(sdmac);
-		else
+		if (sdmac->flags & IMX_DMA_SG_LOOP) {
+			if (sdmac->peripheral_type != IMX_DMATYPE_HDMI)
+				sdma_update_channel_loop(sdmac);
+			else
+				dmaengine_desc_get_callback_invoke(&sdmac->desc,
+									NULL);
+
+		} else
 			tasklet_schedule(&sdmac->tasklet);
 
 		__clear_bit(channel, &stat);
-- 
1.7.9.5

